Gems/PhysX/Code/Editor/Source/ComponentModes/Joints/JointsComponentMode.cpp

```diff
@@ -0,0 +1,598 @@
+        }
+
+        return actions;
+    }
+
+    AZStd::vector<AzToolsFramework::ViewportUi::ClusterId> JointsComponentMode::PopulateViewportUiImpl()
+    {
+        return m_modeSelectionClusterIds;
+    }
+
+    void JointsComponentMode::SetCurrentMode(JointsComponentModeCommon::SubComponentModes::ModeType newMode, ButtonData& buttonData)
+    {
+        auto subMode = m_subModes.find(newMode);
+        if (subMode != m_subModes.end())
+        {
+            const AZ::EntityComponentIdPair entityComponentIdPair = GetEntityComponentIdPair();
+            m_subModes[m_subMode]->Teardown(entityComponentIdPair);
+            m_subMode = newMode;
+            subMode->second->Setup(entityComponentIdPair);
+
+            // if this button is on a different cluster. clear the active state.
+            if (m_activeButton.m_clusterId != buttonData.m_clusterId)
+            {
+                AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(
+                    AzToolsFramework::ViewportUi::DefaultViewportId,
+                    &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::ClearClusterActiveButton, m_activeButton.m_clusterId);
+            }
+            AzToolsFramework::ViewportUi::ViewportUiRequestBus::Event(
+                AzToolsFramework::ViewportUi::DefaultViewportId,
+                &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::SetClusterActiveButton, buttonData.m_clusterId,
+                buttonData.m_buttonId);
+            m_activeButton = buttonData;
+        }
+        AZ_Assert(subMode != m_subModes.end(), "PhysXJoints Uninitialized joint component mode selected.");
+    }
+
+    bool JointsComponentMode::HandleMouseInteraction(const AzToolsFramework::ViewportInteraction::MouseInteractionEvent& mouseInteraction)
+    {
+        // Propagate mouse interaction to sub-component mode.
+        if (m_subModes[m_subMode])
+        {
+            m_subModes[m_subMode]->HandleMouseInteraction(mouseInteraction);
+        }
+
+        return false;
+    }
+
+    void JointsComponentMode::SetupSubModes(const AZ::EntityComponentIdPair& entityComponentIdPair)
+    {
+        //create the 3 cluster groups
+        for (auto& clusterId : m_modeSelectionClusterIds)
+        {
+            AzToolsFramework::ViewportUi::ViewportUiRequestBus::EventResult(
+                clusterId, AzToolsFramework::ViewportUi::DefaultViewportId,
+                &AzToolsFramework::ViewportUi::ViewportUiRequestBus::Events::CreateCluster,
+                AzToolsFramework::ViewportUi::Alignment::TopLeft);
+        }
+
+        //retrieve the enabled sub components from the entity
+        AZStd::vector<JointsComponentModeCommon::SubModeParamaterState> subModesState;
+        EditorJointRequestBus::EventResult(subModesState, entityComponentIdPair, &EditorJointRequests::GetSubComponentModesState);
+
+        const AzToolsFramework::ViewportUi::ClusterId group1ClusterId = GetClusterId(ClusterGroups::Group1);
```

Comment:
question: Is there a more context appropriate name than `Group1/2/3`? Could they refer to the use case or is this meant to be more generic intentionally?

Reply:
i couldn't think of a more specific name TBH and it seemed to fit ok. For context, i'm creating 3 cluster UIs to group different sets of buttons, so thought group 1/2/3 would fit best.

Reply:
Okay got you, makes sense that the groups could contain any set of operations so concrete naming doesn't make sense, thanks for the context